SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c

```diff
@@ -4651,8 +4690,49 @@ SecureBootCallback (
       // When entering SecureBoot OPTION Form
       // always close opened file & free resource
       //
-      if ((QuestionId == KEY_SECURE_BOOT_PK_OPTION) ||
-          (QuestionId == KEY_SECURE_BOOT_KEK_OPTION) ||
+      if ((QuestionId == KEY_SECURE_BOOT_PK_OPTION)){
+        CloseEnrolledFile (Private->FileContext);
+
+        GetVariable2 (EFI_PLATFORM_KEY_NAME, &gEfiGlobalVariableGuid, (VOID **)&PkList, NULL);
+        if (PkList != NULL) {
+          // Extract the PK certificate from the list
+          PkCert = (EFI_SIGNATURE_DATA *)((UINT8 *)PkList + sizeof (EFI_SIGNATURE_LIST) + PkList->SignatureHeaderSize);
+          UINTN cnSize = 0;
+          X509GetCommonName(
+            (UINT8*)PkCert->SignatureData,
+            (UINTN)PkList->SignatureSize,
+            NULL,
+            &cnSize);
+
+          PkCN8 = AllocateZeroPool(cnSize*sizeof(CHAR8));
+          if (PkCN8 == NULL) {
+            return EFI_OUT_OF_RESOURCES;

```

Comment:
Returning here will result in PkList not being freed
Typically this is solved with gotos instead of returns, and a label (e.g. `Exit`) at the end that frees any allocated buffers.

See this for example https://github.com/Dasharo/edk2/blob/0dce26dea89dfd751e024898088a2a2d3f3f9f27/NetworkPkg/IScsiDxe/IScsiDhcp6.c#L456